src/index.ts

```diff
@@ -1,3 +1,182 @@
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 'use-strict';
+
+import {
+  IInputAreaViewModel
+} from 'jupyter-js-input-area';
+
+import {
+  IOutputAreaViewModel
+} from 'jupyter-js-output-area';
+
+import {
+  IObservableList
+} from 'phosphor-observablelist';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+import './index.css';
+
+
+/**
+ * An enum which describes the type of cell.
+ */
+enum CellType {
+  /**
+   * The cell contains code input.
+   */
+  Code,
+
+  /**
+   * The cell contains markdown.
+   */
+  Markdown,
+
+  /**
+   * The cell contains raw text.
+   */
+  Raw
+}
+
+
+/**
+ * An enum which describes which interface property has changed.
+ */
+export
+enum CellProperty {
+  CellType,
+  Tags,
+  Deletable,
+  Mergable,
+  Splittable,
+  Rendered,
+  Marked,
+  Format
+}
+
+
+/**
+ * The arguments object emitted with the `stateChanged` signal.
+ */
+export
+interface ICellChangedArgs<T> {
+  which: CellProperty,
+  oldValue: T;
+  newValue: T;
+}
+
+
+/**
+ * The definition of a model object for a base cell.
+ */
+interface IBaseCellViewModel {
+
+  /**
+   * The type of cell.
+   */
+  cellType: CellType;
+
+  /**
+   * Tags applied to the cell.
+   */
+  tags?: IObservableList<string>;
+
+  /**
+   * Get namespaced metadata about the cell.
+   */
+  getMetadata(namespace: string) : IObservableMap<string, JSONData>;
```

Comment:
> It's the type safety i'm worried about here. It's whether the
> cell/notebook *view* model is the right place for it. Are we expecting
> *all* implementations of ICell to provide this generic (meta)data store?
> If so, are they free to implement it in-memory only?
>
Yes, I would expect all implementations to provide this. I don't see any
problem with in-memory implementations. The only constraint - and this is a
type question - is that at some point, this data will be passed to a
serialization interface, which will attempt to JSON serialize it. This is
where TypeScript's lack of a JSONableData type is limiting. That would be
the the most accurate, and fully typesafe approach for the values in the
store.


> Otherwise, the metadata could be accessible at a lower layer of the model
> hierarchy, and we can expose a way to get to access those models which will
> be available to the cell toolbar.
>
That is an option, but I don't see that approach buying us any additional
type safety - it is just pushing it down one level.



> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/jupyter-js-cells/pull/2/files#r43834178>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com